### DISCUSSION TITLE
How to stop lualine from flickering when selecting options in the coc menu?

### DISCUSSION THREAD

**Discussion by skamsie** [neoclide/coc.nvim#5154] [Q&A] [answered]
I am using [lualine](https://github.com/nvim-lualine/lualine.nvim) and each time I navigate through the coc menu with `Ctrl-P` `Ctrl-N` my status line flickers. I can see it is because with each movement in the coc menu it briefly enters command mode so the status line updates from 'normal' to 'command' and viceversa... Is there a way to prevent this?

my coc settings are minimal:

```lua
return {
  'neoclide/coc.nvim',
  config = function()
    vim.g.coc_global_extensions = {
      'coc-clangd',
      'coc-css',
      'coc-html',
      'coc-lua',
      'coc-solargraph',
      'coc-solargraph',
      'coc-vimlsp',
    }

    vim.cmd[[
      nmap <silent> gd <Plug>(coc-definition)
      nmap <silent> gy <Plug>(coc-type-definition)
      nmap <silent> gi <Plug>(coc-implementation)
      nmap <silent> gr <Plug>(coc-references)
    ]]
  end
}
```

https://github.com/user-attachments/assets/64ebb085-5d47-432d-8368-541a6bb2f80d

**Comment by hexh250786313** [comment]
@skamsie Hi.

```vim
inoremap <silent><expr> <TAB>
      \ coc#pum#visible() ? coc#pum#next(1) :
      \ CheckBackspace() ? "\<Tab>" :
      \ coc#refresh()
inoremap <expr><S-TAB> coc#pum#visible() ? coc#pum#prev(1) : "\<C-h>"
```

I guess your coc configuration code is similar to the above configuration. The issue lies in the `coc#pum#next` and `coc#pum#prev` functions. Let's take a look at their source code:

https://github.com/neoclide/coc.nvim/blob/3b8869a7a40675c49731397f0d1e99e594a9c087/autoload/coc/pum.vim#L134-L136

In fact, they call `\<C-r>=coc#pum#_navigate(1,".a:insert.")\<CR>`, which can be understood as executing `:call xxxxx` in the vim command. Using `:` as the command prefix will cause lualine to flicker. This is not only an issue with coc, but other plugins and even your custom commands may have this problem. Therefore, I recommend using `<cmd>call xxx` to call functions, which will avoid the flickering issue. Of course, I think it would be better for lualine to solve this problem itself.

Back to the coc issue, you can try modifying your configuration by changing `coc#pum#next()` and `coc#pum#prev()` to:

```vim
inoremap <silent><expr> <TAB>
  \ coc#pum#visible() ? "<cmd>call coc#pum#_navigate(1,0)<cr>" :
  \ CheckBackspace() ? "\<Tab>" :
  \ coc#refresh()
inoremap <expr><S-TAB> coc#pum#visible() ? "<cmd>call coc#pum#_navigate(0,0)<cr>" : "\<C-h>"
```

This way, the original functions are called through `<cmd>`, which should avoid the flickering issue. Although this is not the optimal solution, it can temporarily solve your problem.

**Reply by skamsie** [reply]
thanks for your reply, but I already posted my `coc` settings in the original message and I do not have any custom key bindings, I just use `Ctrl-N` and `Ctrl-P`. In any case, the problem seems to be with `lualine` and I opened an issue there https://github.com/nvim-lualine/lualine.nvim/issues/1313

**Reply by hexh250786313** [reply, marked as answer]
@skamsie 

It might be that I didn't explain it clearly enough before. What I mean is that whether it's the default keymap of coc or your custom keymap, as long as it uses the `coc#pum#prev` and `coc#pum#next` keymaps, this issue will occur.

https://github.com/neoclide/coc.nvim/blob/3b8869a7a40675c49731397f0d1e99e594a9c087/plugin/coc.vim#L710-L715

You can see in the coc source code that these two keymaps are bound by default. So, as I mentioned above, you can set a keymap to override these two keymaps, and this issue will not occur. The functionality of these two is exactly the same; the only difference is that the command prefix is changed from `:` to `<cmd>`:

```
inoremap <silent><expr> <C-n> coc#pum#visible() ? "<cmd>call coc#pum#_navigate(1,0)<cr>" : "\<C-n>" 
inoremap <silent><expr> <C-p> coc#pum#visible() ? "<cmd>call coc#pum#_navigate(0,0)<cr>" : "\<C-p>" 
```


https://github.com/user-attachments/assets/2dae0a05-3a97-4c78-9b21-da222cf87d24
